data/transactions/logic/eval.go

```diff
@@ -355,6 +361,10 @@ func EvalStateful(program []byte, params EvalParams) (pass bool, err error) {
 	cx.EvalParams = params
 	cx.runModeFlags = runModeApplication
 	pass, err = eval(program, &cx)
+	if pass && cx.EvalParams.Proto.EnableAppCostPooling {
+		// if eval passes, then budget is always greater than cost, so should not have underflow
+		*cx.EvalParams.PooledApplicationBudget -= cx.cost
+	}
```

Comment:
If an app doesn't pass, everything's probably going to get aborted. Nonetheless, I don't see a good reason not to decrease the pooled budget.

This also needs to be *super* careful about underflow.  I would use SubSaturate or similar.  Going "negative" here almost certainly leads to the possibility of a the budget being interpreted as a high number elsewhere.